docs/design/gitattributes.md

```diff
@@ -0,0 +1,603 @@
+    * Test that a pattern in a subdirectory's `gitattributes` file overrides one in a parent directory.
+    * Test a file in a subdirectory that matches a pattern in a root `gitattributes` file.
+    * Test that if patterns match a directory, files under the directory shouldn't match, e.g., create a file `foo/bar.txt`, and a `gitattirbutes` line `foo text`. Querying the `text` attributes for `foo/bar.txt` should result `Unspecified`.
+    * Test that attributes introduced in a subdirectory are also correctly reported, e.g., we don't create the root `.gitattributes` file, we create the `foo/.gitattributes` file with a `bar.txt text` line, we create the `foo/bar.txt` file, and we query the state of the `text` attribute for the `foo/bar.txt` file.
+
+* Test macro definitions are resolved correctly:
+    * In the `gitattributes` file at the root folder, we define 2 macros:
+    ```gitattr
+    [attr]base_macro a
+    [attr]override_macro b
+    ```
+    * In the `foo1` folder, we create a `bar` file, we define the macro and an attribute assignment to the `bar` file:
+    ```
+    # this macro definition shouldn't take effect.
+    [attr]override_macro c
+    # this macro definition shouldn't take effect.
+    [attr]new_macro e
+    bar base_macro override_macro new_macro
+    ```
+    * In the `foo2` folder, we create a `bar` file as well, and create a similar `.gitattributes` file:
+    ```
+    # this macro definition shouldn't take effect.
+    [attr]override_macro d
+    # this macro definition shouldn't take effect.
+    [attr]new_macro e
+    bar base_macro override_macro new_macro
+    ```
+    * Then we query the `a`, `b`, `c`, `d`, `e`, and `new_macro` attributes for the both file `foo1/bar` and `foo2/bar`. We should expect that `a`, `b`, and `new_macro` are `Set`, while `c`, `d`, and `e` are `Unspecified` for both files.
+
+* Multiple `FileLoader`s and fallback:
+
+    * Using the mocked `FileLoader`, simulate a `gitattributes` file present for the first `FileLoader` but not the second `FileLoader`. Search should use the version form the first `FileLoader`.
+    * Using the mocked `FileLoader`, simulate a `gitattributes` file absent from the first `FileLoader` but present in the second `FileLoader`. Search should use the version form the second `FileLoader`.
+    * Using the mocked `FileLoader`, simulate a `gitattributes` file present in both the first `FileLoader` and the second `FileLoader`. Search should use the version form the first `FileLoader`.
+    * Using the mocked `FileLoader`, simulate a `gitattributes` file absent in both the first `FileLoader` and the second `FileLoader`. Search should result in unspecified for the attributes queried.
+
+* Stress test: create a file path with 200 components, create a file under that path, create a `.gitattributes` file under that path, and verify that querying the attributes defined for the file works without stack overflow.
```

Comment:
optional: Does anyone some repos which we could use to use in a benchmarking suite? So we'd not needlessly run into regressions.

Reply:
We can just use firefox and https://github.com/NixOS/nixpkgs for benchmark, but we do need a performance benchmark monitor platform. Let me ask in the discord channel and come back to see if we need to file a separate issue/design for that.